### DISCUSSION TITLE
Polling onSuccess in version 5

### DISCUSSION THREAD

**Discussion by oren-legit** [TanStack/query#6402] [Q&A] [answered]
When version 5 removed the callbacks, I actually managed to make my code a lot better in most cases.
But there's a case where I can't find a solution besides `useEffect`:

There are specific APIs that take a long time to do something, so we send a POST and start polling on a progress service to see the progress and show it to the user.
When the task is done (the latest polling got a "success" message), we want to invalidate a specific list that is now changed because of the mutation.

What's the right thing to do here?

**Comment by TkDodo** [comment, marked as answer]
useEffect is fine, you could also do in the queryFn of the polling query if you want to avoid the effect
